### ISSUE TITLE
Please don't encourage bad practices

### ISSUE THREAD

**Issue by evanfoster** [cjbassi/gotop#10]
Apologies for this, I know everyone hates it when someone brings this up, but please don't encourage the bad practice of running uninspected, untrusted code off of the internet.

```shell
sh -c "$(curl $link)"
```

This is really, really not good. The user has no chance to inspect the script. Please encourage this practice instead:

```shell
curl -O /tmp/get-gotop.sh && vim /tmp/get-gotop.sh
# Happy with the script?
sh -c /tmp/get-gotop.sh
```

I know it's not as convenient, but it's far safer. Alternatively, you could use the awesome [fpm](https://github.com/jordansissel/fpm) to just make packages for the various distros. `fpm` makes it really simple and pleasant to do so, and this way the user isn't installing things on their system that their package managers can't see.

Again, sorry to bring this up, and I really love the project! I'm going to use this a ton at my job where we have systems with no package manager. I've missed having `htop` and this is an awesome replacement. I just get anal retentive about this particular thing.

**Comment by cjbassi** [comment]
You're definitely right. I'll check out fpm but for now I'll actually just have people download the repo using git since that feels more robust.

**Comment by cjbassi** [comment]
Does this:
```
git clone --depth 1 https://github.com/cjbassi/gotop.git /tmp/gotop
/tmp/gotop/download.sh
```
seem like good practice? Or should I just use what you had above?

**Comment by petermbenjamin** [comment]
@cjbassi - I think your revised approach is good.
The point is to avoid instructing users from running untrusted/unvetted code without a chance for them to review it (either by reviewing the installer script or by reviewing it via their package manager of choice).

[GoReleaser](https://github.com/goreleaser/goreleaser) is a fantastic Go project for building, packaging, and distributing binaries on different platforms as well.

I highly recommend it.

**Comment by cjbassi** [comment]
Thanks for the GoReleaser recommendation, just implemented that into the project to replace the current build process. Gonna learn how to integrate it with nfpm and homebrew which should improve things too.

**Comment by petermbenjamin** [comment]
Not to beat a dead-horse. I just came across this article: https://www.idontplaydarts.com/2016/04/detecting-curl-pipe-bash-server-side/
idontplaydarts.com

TL;DR -
`curl http://<some-site>/script.sh | bash` is detectable server-side.

This means someone could visually inspect `script.sh`, find that it is not harmful, and upon `curl | bash`, the server would detect this and manipulate or change the response to something else.

I thought this article explained the risks well and wanted to share it for edification.

**Comment by cjbassi** [comment]
On second thought, I don't see a problem with curling into bash from a trusted website over https if you've checked the script source on the server. And I've noticed other major projects curling into bash too, like homebrew and rust. I get it's a meme and all, but I'm not sure there's any actual problem here. If people want, I think a good compromise would be to put a disclaimer/warning in the readme about only curling into bash for trusted sources.

**Comment by cjbassi** [comment]
While I think the real world risk of bash curling a script from github is nil, I think you have a point in a general sense, which raises the issue of promoting bad practices for people who don't know what they're doing. Which is where I think a disclaimer would help, unless that's a dumb idea and not a sufficient enough solution to enabling potentially bad practices.
